### ISSUE TITLE
Translatable 'About'

### ISSUE THREAD

**Issue by ghost** [kamilion/customizer#121]
**Last revised: 2016.09.17**

This is probably the most persistent and longest standing issue for Customizer. Until last year, there were only four tabs: "Main", "Extra", "Settings", "About". But that has changed with commit ccaeab6 with attempt to make [`Contributors`](https://github.com/clearkimura/Customizer/blob/master/Contributors) file translatable. Unfortunately, that wasn't the end of story.

**The problem**

It is apparent that `Contributors` file is _not_ GUI file by itself, so it cannot be translated using Qt Linguist. Within the file, the part that need to be translated is just the heading text i.e. "Development Team", "Git commits", etc. The name and e-mail address of contributors are _not_ to be translated for sure.

The `Contributors` file contains Markdown syntax while the file itself is not having file extension of `.md`. Markdown is supposedly useful for simple styling and easier convertion to an HTML file via the `Makefile` or Python scripts (`genabout.py` was created for this).

**Justification**

Do we need the heading text to be translated? Preferably, yes. (Such translation is indeed possible, but we don't know how to implement this properly; Hence, we 'prefer').

To begin with, the contents of `Contributors` and License (Short version of `COPYING`) are hardcoded into GUI file. These have been hardcoded since Gambas ports (see [FAbout.form](https://github.com/clearkimura/Customizer/blob/gambas3/GUI/.src/FAbout.form) and commit cca469c) and even after GUI has been rewritten in Python (see commit 7fc37cc).

Note that `Contributors` that has been converted into HTML syntax is used to hardcoded in GUI file, and _not_ Markdown. This workaround cause duplicated efforts to update the list of Contributors at multiple locations (`Contributor`, `src/gui.ui`) and including translation files (`tr/*.ts`).

This duplicated efforts strongly suggests that we need a permanent solution.

Despite I'm neither a certified nor a competent programmer, I can't help but to run away (laugh at me, if you must) everytime I ran into this issue i.e. whenever updating documentation and list of contributors.

**Study cases**

There are three types of translation methods that can be observed. The following are the translation method in use with real example of application software on Linux platform.
- case [a]: Translatable when run using `env LANG=xx_XX.UTF-8` command workaround in system default locale i.e. some Qt applications such as Customizer GUI, Qt 4 Linguist, WinFF Qt and some GTK+ applications such as Xfce Settings Manager, but not applicable to VLC;
- case [b] Translatable when run in target locale (changed from system default) i.e. some GTK+ applications such as About Xfce, Thunar, GNOME Disks and some applications in case [a];
- case [c] Translatable via settings (limited within application itself) i.e. Supertuxkart and some other applications, but not applicable to above;

Qt framework makes possible for Qt application GUI to be translated for both cases, [a] and [b], by making use of Qt translation files (`*.tr`) and its binary files (`*.qm`). But for elements that are _not_ part of GUI, these elements cannot be translated via Qt framework.

Case [c] is relatively easy to implement (I think so) and not necessarily bad. However, end users tend to look for "out of box" experience, such that applications already been translated when the application is running in target locale. For this reason, method used in case [c] shall be avoided. 

**Implementation stages**
1. Decide the method to translate the `Contributors` file
   
   -- proposed approach: Similar to GUI translation, generate translated HTML files via `Makefile` and put these files in `tr/` directory, then include the files in `resources.qrc` with languages noted, use QTranslator() to handle locale as hinted in [this thread on Qt Centre forum](http://www.qtcentre.org/threads/61442-What-is-needed-to-set-language-for-resource-%28qrc%29-to-use) (Untested)
2. Decide the appearance of GUI for "About" tab
   
   -- proposed approach: Use [the modified `gui.ui`](https://github.com/clearkimura/Customizer/blob/translatable-about/src/gui.ui) from [**translatable-about** branch](https://github.com/clearkimura/Customizer/tree/translatable-about). Then, in "About" tab, replace QTextEdit with QStackedWidget to handle multiple pages, as hinted in [this post on Stack Overflow](http://stackoverflow.com/q/14060703). This way, the same widget area can be reused, and no pop-up window or extra tab is needed to display the contents (Untested)
   
   -- in addition to existing "License" button, include two more buttons to replace the horizontal spacers;
   
   -- on the left, add second button "More about" that displays the logo (full size at 128 x 128 pixels) in left column and few lines of text about Customizer (detailed version by commit number and commit date) in right column;
   
   -- on the right, add third button for "Credits" that displays the HTML file using QTextBrowser (not QTextEdit) because it can specify path to target file using resource browser.
   
   -- similarly, to display content of `License`, use QTextBrowser (not QTextEdit); Except, the content is provided by `License.html`. Note that this static HTML doesn't need to be translated because [GNU GPL v2.0 translations are unofficial](https://www.gnu.org/licenses/old-licenses/gpl-2.0-translations.en.html)
   
   -- clicking on each button will change the display of pages, which are contained in QStackedWidget; This shall be implemented similar to functions of existing buttons in `gui.py.in` (Untested)
3. Implement changes to GUI and internal scripts separately
   
   -- with respect to 1 and 2, GUI is updated first, regardless of translation method is ready or not;
   
   -- the translation method is implemented by scripts that only affect internal working of `Makefile` and Python scripts; The scripts' role is to translate HTML files from the source `Contributors` file, while GUI file holds path to these files, which is set from Resource browser in Qt Designer.
   
   -- when making changes to GUI, remove the hardcoded parts first. Then, display default files (English) i.e. `Contributors.html` (existing file with Markdown syntax that is converted using `python -m markdown INPUTFILE > OUTPUTFILE` command, whose module provided by `python-markdown` for Python 2.x versions) and `License.html` (Create this file to contain the hardcoded text in `gui.py.in`)

**Alternative approach**

Forget about above implementation that requires Markdown-to-HTML conversion!

We think too hard on how to translate the existing document "as it is", because we have already seen other projects i.e. Xfce that such translation is possible. But those projects use different development toolkit and use advanced tools to handle the translation, which is quite wasteful or extravagant for smaller community projects such as Customizer.

I have thought that solving this issue would be much easier, if the current formatting of `Contributors` file is changed to something more simple. I have thought an alternative approach is to remove the headings and replace with few "tags" instead. The "tags" are non-translatable strings that is included in `Contributors` file. The "tags" and its description are labels that include as part of GUI, which makes possible for the description (as replacement for heading) to be translatable.

UI sketch will be posted as new comment at below.

**Remarks**

The implementation stages are subject to changes and depending on the outcome of discussion with the original developer. This would be the second attempt to discuss this issue.

Sounds easy to say than actually done. I probably need to do more trials and errors to prove that the proposed approaches actually work, so I may _not_ discuss this issue without any testing result.

**Comment by ghost** [comment]
**TL;DR for implementing translatable 'About'**
- For GUI, use QStackWidget, QTextBrowser (display credits, licenses, logo page) and QLabel (display 128 x 128 icon in logo page) to implement the changes
- For internal scripts, requires additional dependency `python-markdown` package to convert Markdown text files to HTML files that can be displayed by QTextBrowser via resource browser in Qt Designer (or `resources.qrc`).
- `Contributors.html`: Translate to other locales and save translated HTML files to `tr/Contributors_LOCALE.html`
- `License.html`: English only (License translations are unofficial/invalid for legal reason)

@fluxer I did some research and I need your opinion regarding the proposed approach. You may leave a comment below as usual or, send an email for closed discussion (whichever you prefer). I need to cool down for several days because too much reading; I hope to discuss this issue 'slowly'.

**Comment by fluxer** [comment]
@clearkimura If you insist on fixing this then I can only suggest you to make a quick and simple solution, you or the person who is going to work this out. That is the text that must be translated (contributors and copying, at your choice/not necessary both) should be part of the .py/.ui files of the project. Then figure out how to make it all work out with something like Transifex. If the translation process is not easy and straightforward many people/translators will not even attempt to make translations, as has time proven despite the support for internationalization of the project.

**Comment by ghost** [comment]
**TL;DR for implementing alternative approach**

I hope to simplify the way contributors are listed, so that translation is made easier.
- New formatting will be written in plain text, without Markdown styling at all;
- Use one line per contributor, arranged in alphabetical order; Include only most important information: full name, nickname, contribution type;
- No more headings and redundant text (I will move some text `If you know of anyone missing from this list...` to README.md, which will be revised later in patch release 4.1.?);
- Use tag to denote respective type of contribution i.e. `D` (tag) for `Documentation` (description), etc; Both tag and description are GUI labels, but only description is required to be translatable;
- A contributor can have multiple tags, which are also listed in alphabetical order; 

To illustrate this approach, I have created a sketchy UI as below:

![ctmz-gui-about-credits](https://cloud.githubusercontent.com/assets/1223304/18610345/c0a9e36e-7d4c-11e6-8068-65fe73cbdadc.png)

Note that all contributors will have at least `C` tag, which denote general contributions of one or more kinds: feedback issues, bug reports, pull requests and other noteworthy efforts.

I may add one more tag `R` (missing in above sketchy UI) to denote tools and resources used by Customizer, which are subject to CC license that requires proper attribution. 

@fluxer This approach is much simple and make use of existing Qt translation tool. Listing will be more fair and neutral. This is the easiest way I could think of. What do you think?

**Comment by fluxer** [comment]
@clearkimura I think this is the first time I see something like this - tagging what kind of contribution is made. I'm sure KDE has something similar where the developers can add additional text bellow the contributor name but it's far more complicated, even with optional buttons for e-mail and web address. In this case - I'm not sure the tags are needed. I'm not saying they are bad idea but maybe you should check out how other projects have done their about dialog/tab/etc. and make it a middle ground solution as some people like unified look and such.

Personally I do not care what it looks like anymore, as long as the contributors list is up-to-date and it's easy to do so I'm happy. The only problem I have with is the fact that I have missed to keep it update it every time someone makes a contribution - there are people who have been there for me when I needed them but sadly are not in the list and I sincerely regret about not noting that.

On the upside - nowdays thanks to services like GitHub history is unlikely to get lost. Issues, pull requests, etc. - it's all there. For an example the (somewhat accurate) contributors list is available via https://github.com/clearkimura/Customizer/graphs/contributors. I know I'm kinda stretching the issue here but I'm gonna ask a question - do we need this kind of thing anymore? Maybe we can just leave it as it is, put a link to a page on GitHub somewhere in the 'About' and just make use of the solutions available. There is a negative side when using the service because it requires Internet connection, the contributors names are not translatable, some contributions are not listed there because they are pure suggestions, etc. but all-in-all - I do not think it is worth it to invest much time into this. If someone wants to make a suggestions, fix a bug or something else and be noted as contributor - there is a way. We cannot compensate for all deficiencies, we can only make some of them less painful. There are far more important things that must be done IMO to the project, like porting it to PyQt5.

Cheers!

**Comment by ghost** [comment]
**Implementing without translatable 'About'**

To avoid all the hassles of translation, I wish we could simplify the listing to just contributor fullname and nickname (username) only. Then, GUI will be at simplest as below:

![ctmz-gui-about-credits-2](https://cloud.githubusercontent.com/assets/1223304/18614531/9aad83fc-7dc2-11e6-9d5a-4c64816fd9c3.png)

@fluxer Replied as below

> [...] maybe you should check out how other projects have done their about dialog/tab/etc. [...] 

Regarding credits or contribution records, I have checked out many projects already (See "Study cases" in my top comment). Most projects will do the "detailed" listing that includes type of contribution. I think we both agree that we want to avoid the "detailed" listing, once and for all.

> Personally I do not care what it looks like anymore, as long as the contributors list is up-to-date and it's easy to do so I'm happy. 

The simplest way is to list down the full name only, for example, [contributors of Mozilla project](https://www.mozilla.org/credits/). But they require contributors themselves to fill in a form with proof of contribution (I think we don't need this). In our case, if someone contributed to our project, we just add to our listing "as we notice", else someone can create new issue to notify us (I can create `ISSUE_TEMPLATE.md` to remind users for this).

> The only problem I have with is the fact that I have missed to keep it update it every time someone makes a contribution 

That is because our contributors listing is... too detailed and difficult to update. Even I have made mistake before, when deciding where to put the contributors' name, until you reminded me "everyone are contributors". I'm still ashamed of my mistake.

> I know I'm kinda stretching the issue here but I'm gonna ask a question - do we need this kind of thing anymore?

If you mean "this kind" by mentioning contribution details, I'd say "No, we don't need details."

If you mean "this kind" by listing contributors' name, I'd say "Yes, we shall have the contributors' names only". That is the simplest way to give credit without translation hassles.

>  I do not think it is worth it to invest much time into this.

I agree. That's why I have suggested to simplify our contributors listing. No more translation hassles. No more markdown styling. No more details. Just the contributors' names in plain text.

> There are far more important things that must be done IMO to the project, like porting it to PyQt5.

Each contributor has different role with different priority. For my role, as a long time documentation contributor, doing documentation clean up and citing proper attribution are more important.

Strangely, what I am doing now is beyond that -- reviewing source code, helping out to clean up GUI appearance and try to simplify things. I just wanted to clean up things so that I hope this project would attract future contributors, who will less likely to see the mess we have made.

**TL;DR** I will be implementing in stages and changes to GUI will be included later in future release 4.2.0. Shall I proceed with the simplified contributors listing and simplified GUI?

**Comment by fluxer** [comment]
@clearkimura 

> I have checked out many projects already

I know you did you homework ;). Like you looked into the contributors of Mozilla project, but that's way too complicated and I do not want to force forms onto every contributor. I was just noting in case you really want to invest even more time into this.

> The simplest way is to list down the full name only.

Agree. While having e-mails in the contributors list is desirable there are many possibilities to abuse the information.

> we just add to our listing "as we notice", else someone can create new issue to notify us

This is what we have done so far but since we are slowly stepping away from the project we may have to take a different approach to this. Or not and stick with a simple list.

> That is because our contributors listing is... too detailed and difficult to update

No, that's because the project started years ago and I was not aware of the usual practices in software development. You have done an amazing job at making the list as complete as possible, remember our discussion via e-mail?

> If you mean "this kind" by mentioning contribution details, I'd say "No, we don't need details."
> 
> If you mean "this kind" by listing contributors' name, I'd say "Yes, we shall have the contributors' names only". That is the simplest way to give credit without translation hassles.

I mean a list of contributors in any form that we have to update.

> I agree. That's why I have suggested to simplify our contributors listing. No more translation hassles. No more markdown styling. No more details. Just the contributors' names in plain text.

Sound good, to me. Let's see if someone else wants to have a word on the matter.

> Each contributor has different role with different priority. For my role, as a long time documentation contributor, doing documentation clean up and citing proper attribution are more important.

I know. And you are doing fantastic at that. It's just that the world does not stop to spin just because we are being sidetracked from the project, Qt4 is being obsoleted and soon will not be available in many distributions. [BLFS](http://www.linuxfromscratch.org/blfs/) (it's a book but still), [KaOS](https://kaosx.us/), etc. are all moving away from it, just like with the i386 architecture.

> Strangely, what I am doing now is beyond that.. . I just wanted to clean up things so that I hope this project would attract future contributors, who will less likely to see the mess we have made.

I noticed that. Sadly, I'm no longer motivated to work on the project. With systemd/upstart, Mir/Wayland/X11, etc. and all the possible horrible and incomplete solutions out there, there is a ton of work that needs to be done to keep it up with what is coming and I just feel like any effort to clean things up and improve the project is a waste of time and energy. But that's just me.

> I will be implementing in stages and changes to GUI will be included later in future release 4.2.0. Shall I proceed with the simplified contributors listing and simplified GUI?

Sure. :+1:

**Comment by ghost** [comment]
The 'About' tab has been restored and hardcoded strings were removed. The 'About' tab no longer need to be translated. Resolved in last update for "old stable".
